ops/relation.py

```diff
@@ -0,0 +1,548 @@
+
+    def ready(self):
+        """Set provider state to ready
+        """
+        if not self.is_ready:
+            logger.debug("Provider is ready")
+            self.stored.ready = True
+            self.notify_consumers()
+
+    def unready(self):
+        """Set provider state to unready
+        """
+        logger.debug("Provider is not ready")
+        self.stored.ready = False
+        self.notify_consumers()
+
+    def provider_data(self):
+        """Construct relation data packet for consumer
+        """
+        data = self.provides.copy()
+        data['ready'] = self.stored.ready
+        return data
+
+    @property
+    def is_ready(self):
+        """Query state of provider
+        """
+        return self.stored.ready
+
+
+class ProviderAvailable(EventBase):
+    """Event triggered when a valid provider is found
+
+    When a consumer charm forms a relation with a provider charm,
+    their :class:`Consumer` and :class:`Provider` object exchange
+    information to asertain compatibility. If they relation is found
+    to be compatible then the :class:`Consumer` object raises a
+    :class:`ProviderAvailable` event to inform the consumer charm, that
+    a relation with the provider charm has been successful.
+    """
+    def __init__(self, handle, data=None):
+        super().__init__(handle)
+        self.data = data
+
+    def snapshot(self):
+        return {"data": self.data}
+
+    def restore(self, snapshot):
+        self.data = snapshot["data"]
+
+
+class ProviderInvalid(EventBase):
+    """Event triggered when a provider is not compatible
+
+    When a consumer charm forms a relation with a provider charm,
+    their :class:`Consumer` and :class:`Provider` object exchange
+    information to asertain compatibility. If they relation is found
+    not to be compatible then the :class:`Consumer` object raises a
+    :class:`ProviderInvalid` event to inform the consumer charm, that
+    a relation with the provider charm has *not* been successful.
+    """
+    def __init__(self, handle, data=None):
+        super().__init__(handle)
+        self.data = data
+
+    def snapshot(self):
+        return {"data": self.data}
+
+    def restore(self, snapshot):
+        self.data = snapshot["data"]
+
+
+class ProviderUnready(EventBase):
+    """Event triggered when a provider is not ready
+
+    If a provider charm is not ready to service requests, when a
+    consumer charm for a new relation with it, or is already related
+    to it, then a :class:`ProviderUnready` event is raised. This
+    presumes that the provider charm as set its `ready` status to
+    `False`.
+
+    The :class:`ProviderUnready` event is raised regardless of wether
+    the provider charm is compatible or not. Compatibility checks are
+    done only if the provider charm is ready to service requests.
+    """
+    pass
+
+
+class ProviderBroken(EventBase):
+    """Event raised when provider consumer relation is disolved
```

Comment:
dissolved?

Reply:
Thanks. Should have run a spell check. Will do, fix and push.

Reply:
Fixed. Complete file was spell checked.